File: fluentcore/src/main/java/com/mcreater/fxfluent/stage/FluentStage.java
package com.mcreater.fxfluent.stage;

import com.mcreater.fxfluent.controls.value.AnimatedValue;
import com.mcreater.fxfluent.stage.controls.FluentTitleBar;
import com.mcreater.fxfluent.syslib.UiShellWrapper;
import com.mcreater.fxfluent.util.NativeUtil;
import com.mcreater.fxfluent.util.listeners.NewValueListener;
import com.mcreater.fxfluent.xaml.style.AppColorTheme;
import com.mcreater.fxfluent.xaml.style.SystemThemeLoop;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.Scene;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;
import javafx.stage.Stage;
import javafx.stage.StageStyle;
import javafx.util.Duration;

public class FluentStage extends Stage {
    private UiShellWrapper.BackdropType backdropType;
    private Node content;
    private Pane sceneContent;
    private boolean lastApply = false;
    private boolean isDarkMode = UiShellWrapper.GetSystemIsDark();
    private boolean isFirstInit = true;
    private boolean disableBackdrop = false;
    private boolean disableBackground = false;
    private AppColorTheme colorThemeOverride = AppColorTheme.SYSTEM;

    public void setColorThemeOverride(AppColorTheme colorThemeOverride) {
        this.colorThemeOverride = colorThemeOverride;
        this.applyBackdropType();
    }

    public void setDisableBackdrop(boolean disableBackdrop) {
        this.disableBackdrop = disableBackdrop;
    }

    public void setDisableBackground(boolean disableBackground) {
        this.disableBackground = disableBackground;
    }

    private final AnimatedValue<Color> windowColor = new AnimatedValue<>(Color.rgb(243, 243, 243), Duration.millis(150));
    public FluentStage() {
        this(StageStyle.DECORATED);
    }
    public FluentStage(StageStyle stageStyle) {
        super(stageStyle);
        init();
    }
    private void init() {
        this.backdropType = UiShellWrapper.BackdropType.getDefault();
        this.content = new Pane();
        this.sceneContent = new VBox();
        SystemThemeLoop.addListener(a -> this.applyBackdropType());
    }

    /**
     * Set backdrop type of this stage (Windows only)<br>设置该窗口的背景类型 (仅 Windows)
     * @param backdropType {@link UiShellWrapper.BackdropType}
     */
    public void setBackdropType(UiShellWrapper.BackdropType backdropType) {
        this.backdropType = backdropType;
    }

    /**
     * Apply backdrop type to this stage<br>向该窗口应用背景类型
     */
    public void applyBackdropType() {
        try {
            isDarkMode = colorThemeOverride == AppColorTheme.SYSTEM ? UiShellWrapper.GetSystemIsDark() : (colorThemeOverride == AppColorTheme.DARK);
            if (!disableBackdrop) {
                lastApply = UiShellWrapper.ApplyBlur(NativeUtil.getWindowHandle(this), backdropType, isDarkMode);
            }
            this.updateScene();
            if (!isFirstInit) {
                setIconified(true);
                setIconified(false);
            }
            else isFirstInit = false;
        }
        catch (Exception e) {
            e.printStackTrace();
            UiShellWrapper.NativeWarn();
            lastApply = false;
        }
    }

    /**
     * Set content of this stage<br>设置该窗口的内容
     * @param content {@link Node}
     */
    public void setContent(Node content) {
        this.content = content;
        this.updateScene();
    }

    private FluentTitleBar buildTitleBar() {
        return new FluentTitleBar(this);
    }

    private Color getWindowBackground() {
        if (!disableBackground) {
            int i = isDarkMode ? 32 : 243;
            if (getStyle() != StageStyle.TRANSPARENT) {
                return Color.rgb(i, i, i, 1);
            } else {
                return Color.rgb(i, i, i, (lastApply) ? (UiShellWrapper.needBackground(backdropType) ? .65 : 0) : 1);
            }
        }
        else return Color.TRANSPARENT;
    }

    private void updateScene() {
        int i = isDarkMode ? 32 : 243;
        windowColor.updateValue(getWindowBackground());
        windowColor.property.addListener((observableValue, color, t1) -> FluentStage.this.sceneContent.setBackground(new Background(
                new BackgroundFill(
                        getWindowBackground(),
                        CornerRadii.EMPTY,
                        Insets.EMPTY
                )
        )));
        FluentTitleBar titleBar = buildTitleBar();

        Region placeHolder = new Region();
        placeHolder.setPrefHeight(50 + 10);

        this.sceneContent = new Pane();
        this.sceneContent.getChildren().clear();
        this.sceneContent.getChildren().addAll(new VBox(placeHolder, this.content), titleBar);
        this.sceneContent.setBackground(new Background(
                new BackgroundFill(
                        getWindowBackground(),
                        CornerRadii.EMPTY,
                        Insets.EMPTY
                )
        ));

        Rectangle rectangle = new Rectangle();
        rectangle.widthProperty().bind(this.widthProperty());
        rectangle.heightProperty().bind(this.heightProperty());
        rectangle.setArcWidth(15);
        rectangle.setArcHeight(15);
        maximizedProperty().addListener((NewValueListener<Boolean>) t1 -> {
            rectangle.setArcWidth(t1 ? 0 : 15);
            rectangle.setArcHeight(t1 ? 0 : 15);
        });
        this.sceneContent.setClip(rectangle);
        titleBar.prefWidthProperty().bind(this.widthProperty());

        this.sceneContent.prefWidthProperty().bind(this.widthProperty());
        this.sceneContent.prefHeightProperty().bind(this.heightProperty());

        Scene scene = new Scene(this.sceneContent);
        scene.setFill(Color.TRANSPARENT);
        this.setScene(scene);
    }

   public static class WindowMovement {
        double x1;
        double y1;
        double x_stage;
        double y_stage;

        public static WindowMovement getInstance() {
            return new WindowMovement();
        }

        private WindowMovement() {
        }

        public <V extends Region, K extends Stage> void windowMove(V listenedObject, K stage) {
            listenedObject.setOnMouseDragged(event -> {
                stage.setX(this.x_stage + event.getScreenX() - this.x1);
                stage.setY(this.y_stage + event.getScreenY() - this.y1);
            });
            listenedObject.setOnDragEntered(null);
            listenedObject.setOnMousePressed(event -> {
                this.x1 = event.getScreenX();
                this.y1 = event.getScreenY();
                this.x_stage = stage.getX();
                this.y_stage = stage.getY();
            });
        }
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/stage/FluentStage.kt
package com.mcreater.fxfluent.stage

import com.mcreater.fxfluent.controls.abstractions.SystemThemeListenable
import com.mcreater.fxfluent.controls.value.AnimatedValue
import com.mcreater.fxfluent.stage.controls.FluentTitleBar
import com.mcreater.fxfluent.syslib.BackdropType
import com.mcreater.fxfluent.syslib.BackdropType.Companion.default
import com.mcreater.fxfluent.syslib.UiShellWrapper.Companion.applyBlur
import com.mcreater.fxfluent.syslib.UiShellWrapper.Companion.getSystemIsDark
import com.mcreater.fxfluent.syslib.UiShellWrapper.Companion.nativeWarn
import com.mcreater.fxfluent.syslib.UiShellWrapper.Companion.needBackground
import com.mcreater.fxfluent.util.NativeUtil.Companion.getWindowHandle
import com.mcreater.fxfluent.util.listeners.NewValueListener
import com.mcreater.fxfluent.xaml.style.AppColorTheme
import com.mcreater.fxfluent.xaml.style.SystemThemeLoop
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.scene.Cursor
import javafx.scene.Node
import javafx.scene.Scene
import javafx.scene.input.MouseEvent
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.shape.Rectangle
import javafx.stage.Stage
import javafx.stage.StageStyle
import javafx.util.Duration

class FluentStage(style: StageStyle) : Stage(style) {
    private var isBottomLeft = false
    private var isRight = false
    private var isBottomRight = false
    private var isBottom = false
    private var isTopRight = false
    var maximizedover = false
    init {
        init()
    }
    constructor() : this(StageStyle.TRANSPARENT) {
        init()
    }

    private var content: Node? = null
    private var sceneContent: Pane? = null
    private var lastApply = false
    private var isDarkMode = getSystemIsDark()
    var backdropType: BackdropType? = null
    var disableBackdrop = false
    var disableBackground = false
    private var fluentTitleBar: FluentTitleBar? = null
    var colorThemeOverride = AppColorTheme.SYSTEM
        /**
         * Set backdrop type of this stage (Windows only)<br></br>设置该窗口的背景类型 (仅 Windows)
         * @param [BackdropType]
         */
        set(a) {
            field = a
            applyBackdropType()
        }

    private val windowColor = AnimatedValue(Color.rgb(243, 243, 243), Duration.millis(150.0))
    private fun init() {
        backdropType = default
        content = Pane()
        sceneContent = VBox()
        SystemThemeLoop.addListener { applyBackdropType() }
    }

    /**
     * Apply backdrop type to this stage<br></br>向该窗口应用背景类型
     */
    fun applyBackdropType() {
        try {
            isDarkMode =
                if (colorThemeOverride == AppColorTheme.SYSTEM) getSystemIsDark() else colorThemeOverride == AppColorTheme.DARK
            if (!disableBackdrop) {
                lastApply = applyBlur(
                    getWindowHandle(this),
                    backdropType!!, isDarkMode
                )
            }
            updateScene()
        } catch (e: Exception) {
            e.printStackTrace()
            nativeWarn()
            lastApply = false
        }
    }

    /**
     * Set content of this stage<br></br>设置该窗口的内容
     * @param content [Node]
     */
    fun setContent(content: Node?) {
        if (content is SystemThemeListenable) {
            this.content = content
            updateScene()
        }
    }

    private fun buildTitleBar(): FluentTitleBar {
        return FluentTitleBar(this)
    }

    private fun getWindowBackground(): Color {
        return if (!disableBackground) {
            val i = if (isDarkMode) 32 else 243
            if (style != StageStyle.TRANSPARENT) {
                Color.rgb(i, i, i, 1.0)
            } else {
                Color.rgb(i, i, i, if (lastApply) (if (needBackground(backdropType!!)) .65 else 0.0) else 1.0)
            }
        } else Color.TRANSPARENT
    }

    private fun updateScene() {
        windowColor.updateValue(getWindowBackground())
        windowColor.property.addListener { _: ObservableValue<out Color>?, _: Color?, _: Color? ->
            this@FluentStage.sceneContent?.background = Background(
                BackgroundFill(
                    getWindowBackground(),
                    CornerRadii.EMPTY,
                    Insets.EMPTY
                )
            )
        }
        val titleBar = buildTitleBar()
        fluentTitleBar = titleBar
        val placeHolder = Region()
        placeHolder.prefHeight = (50 + 10).toDouble()
        sceneContent = Pane()
        sceneContent!!.children.clear()
        if (style != StageStyle.DECORATED || style != StageStyle.UTILITY || style != StageStyle.UNIFIED) sceneContent!!.children.addAll(VBox(placeHolder, content), titleBar)
        else sceneContent!!.children.addAll(content)

        (content as Region).prefWidthProperty().bind(widthProperty())
        (content as Region).prefHeightProperty().bind(heightProperty().add(
            if (style != StageStyle.DECORATED || style != StageStyle.UTILITY || style != StageStyle.UNIFIED) -60 else 0
        ))

        sceneContent!!.background = Background(
            BackgroundFill(
                getWindowBackground(),
                CornerRadii.EMPTY,
                Insets.EMPTY
            )
        )
        /*sceneContent?.border = Border(
            BorderStroke(
                Color.rgb(0, 0, 0, 0.5),
                BorderStrokeStyle.SOLID,
                CornerRadii(5.0),
                BorderWidths(1.0),
                Insets.EMPTY
            )
        )*/
        val rectangle = Rectangle()
        rectangle.widthProperty().bind(widthProperty())
        rectangle.heightProperty().bind(heightProperty())
        rectangle.arcWidth = 15.0
        rectangle.arcHeight = 15.0
        maximizedProperty().addListener (NewValueListener {
            rectangle.arcWidth = (if (isMaximized) 0 else 15).toDouble()
            rectangle.arcHeight = (if (isMaximized) 0 else 15).toDouble()
        })
        sceneContent!!.clip = rectangle
        titleBar.prefWidthProperty().bind(widthProperty())
        /*sceneContent!!.prefWidthProperty().bind(widthProperty())
        sceneContent!!.prefHeightProperty().bind(heightProperty())*/

        sceneContent!!.setOnMouseMoved { event: MouseEvent ->
            event.consume()
            val x = event.sceneX
            val y = event.sceneY
            val width: Double = this.width
            val height: Double = this.height
            var cursorType: Cursor = Cursor.DEFAULT
            isBottom = false
            isBottomRight = false
            isRight = false
            isBottomLeft = false
            isTopRight = false
            if (y >= height - 5) {
                if (x <= 5) {
                    isBottomLeft = true
                    cursorType = Cursor.SW_RESIZE
                }
                else if (x >= width - 5) {
                    isBottomRight = true
                    cursorType = Cursor.SE_RESIZE
                }
                else {
                    isBottom = true
                    cursorType = Cursor.S_RESIZE
                }
            }
            else if (y <= 5) {
                if (x >= width - 5) {
                    isTopRight = true
                    cursorType = Cursor.NE_RESIZE
                }
            }
            else {
                if (x >= width - 5) {
                    isRight = true
                    cursorType = Cursor.E_RESIZE
                }
            }
            sceneContent!!.cursor = cursorType
        }
        sceneContent!!.setOnMouseDragged { event: MouseEvent ->
            event.consume()
            if (!maximizedover) {
                val righted = isTopRight || isRight || isBottomRight
                val bottomed = isBottomLeft || isBottom || isBottomRight

                if (bottomed && event.sceneY >= minHeight && event.sceneY >= 60) this.height = event.sceneY
                if (righted && event.sceneX >= minWidth) this.width = event.sceneX
            }
        }

        val scene = Scene(sceneContent)
        scene.fill = Color.TRANSPARENT
        this.scene = scene
    }


    class WindowMovement private constructor() {
        private var x1 = 0.0
        private var y1 = 0.0
        private var x_stage = 0.0
        private var y_stage = 0.0
        fun <V : Region?, K : FluentStage?> windowMove(listenedObject: V, stage: K) {
            listenedObject!!.onMouseDragged = EventHandler { event: MouseEvent ->
                if (stage!!.maximizedover) {
                    stage.fluentTitleBar?.buttonmax?.onAction?.handle(ActionEvent())
                    return@EventHandler
                }
                stage.x = x_stage + event.screenX - x1
                stage.y = y_stage + event.screenY - y1
            }
            listenedObject.onDragEntered = null
            listenedObject.onMousePressed = EventHandler { event: MouseEvent ->
                if (stage!!.maximizedover) {
                    stage.fluentTitleBar?.buttonmax?.onAction?.handle(ActionEvent())
                    return@EventHandler
                }
                x1 = event.screenX
                y1 = event.screenY
                x_stage = stage.x
                y_stage = stage.y
            }
        }

        companion object {
            @JvmStatic
            val instance: WindowMovement
                get() = WindowMovement()
        }
    }

}